### PR TITLE
Added `discretise()` and `parameters()` methods to `<epidist>`

### PR DESCRIPTION
**Pull request by joshwlambert**
This pull request addresses issues #124 & #127. Two new function are added:
- `discretise()` 
- `parameters()`

`discretise()` takes an `<epidist>` object with a continuous distribution (e.g. gamma or lognormal) and converts it to its discretised form. This can be useful when an entry in the database is continuous but a discretised distribution is required for modelling. This function requires adding a new generic.

`parameters()` is a method that extracts the parameters of the probability distribution and returns them in a consistent manner (named vector). This is useful as the functions and accessors differ depending on whether the distribution is discretised (using {distcrete}) or not (using {distributional}), and now the user does not have to remember these differences when working with `<epidist>` objects. This function does not require adding a new generic as it imports the generic from the {distributional} package.

Tests and documentation are updated in accordance with the above-mentioned changes. Other {epiparameter} functions have been updated to use the newly added functions where necessary.

### CODE REVIEW

File: R/epidist.R [hunk 1]
```diff
@@ -429,7 +429,7 @@ format.epidist <- function(x, header = TRUE, vb = NULL, ...) {
   } else {
 
     # decide on parameter format from magnitude of number
-    params <- unlist(distributional::parameters(x$prob_dist))
+    params <- parameters(x)
     format_params <- ifelse(
       test = any(params > 9.999e-3 & params < 1e4),
```
Comment (adamkucharski) [inline R/epidist.R:434]: Presumably this won't break anywhere (e.g. if superspreading has very large MLE for k)?

File: R/epidist.R [hunk 2]
```diff
@@ -704,3 +704,129 @@ generate.epidist <- function(x, times, ...) {
...
+#'   prob_distribution_params = c(shape = 1, scale = 1)
+#' )
+#' discretise(ebola_incubation)
+discretise <- function(x, ...) {
+  UseMethod("discretise")
+}
+
+#' @rdname discretise
+#' @export
+discretise.epidist <- function(x, ...) {
```
Comment (adamkucharski) [inline R/epidist.R:734]: Worth flagging in function description (or somewhere else) that distributional functions may no longer apply to distcrete objects (e.g. `generate(onset_to_death_ebola$prob_dist,10)` vs
`generate(epiparameter::discretise(onset_to_death_ebola)$prob_dist,10)`? Not sure what relevant syntax is for same random draw operation with `distcrete` (and corresponding pdf/pmf functions).

**Review by adamkucharski — COMMENTED** [review]
Have added a couple of comments about potential for things to break in downstream use. If not already planned, would also be useful to update epiparameter vignette to give some easy quick start examples for how to output random draws, density functions (rnorm, dnorm etc.) with these classes.

File: R/epidist.R (hunk 1 above)
Comment (joshwlambert) [reply R/epidist.R:434]: It should work fine. Here's an example of an `<epidist>` object for superspreading (negative binomial) with large and small values for k.

``` r
library(epiparameter)
transmission <- epidist(
  disease = "SARS", 
  epi_dist = "offspring_distribution", 
  prob_distribution = "nbinom", 
  prob_distribution_params = c(mean = 1, dispersion = 1e10)
)
#> Citation cannot be created as either author, year or DOI is missing
transmission
#> Disease: SARS
#> Pathogen: NA
#> Epi Distribution: offspring distribution
#> Study: No citation available
#> Distribution: negbin
#> Parameters:
#>   n: 10000000000.000
#>   p: 1.000

transmission <- epidist(
  disease = "SARS", 
  epi_dist = "offspring_distribution", 
  prob_distribution = "nbinom", 
  prob_distribution_params = c(mean = 1, dispersion = 1e-10)
)
#> Citation cannot be created as either author, year or DOI is missing
transmission
#> Disease: SARS
#> Pathogen: NA
#> Epi Distribution: offspring distribution
#> Study: No citation available
#> Distribution: negbin
#> Parameters:
#>   n: 1e-10
#>   p: 1e-10
```

<sup>Created on 2023-04-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/epidist.R (hunk 2 above)
Comment (joshwlambert) [reply R/epidist.R:734]: The distribution functions in {epiparameter} are setup to handle both discretised and non-discretised distributions:

``` r
library(epiparameter)
library(distributional)
ebola_si <- epidist(
  disease = "ebola", 
  epi_dist = "serial_interval", 
  prob_distribution = "gamma", 
  prob_distribution_params = c(shape = 1, scale = 1)
)
#> Citation cannot be created as either author, year or DOI is missing
ebola_si
#> Disease: ebola
#> Pathogen: NA
#> Epi Distribution: serial interval
#> Study: No citation available
#> Distribution: gamma
#> Parameters:
#>   shape: 1.000
#>   rate: 1.000

cdf(ebola_si, q = 50)
#> [1] 1
generate(ebola_si, times = 10)
#>  [1] 0.2634135 1.1430287 1.0172865 0.4948622 1.4337434 0.8281407 0.5926481
#>  [8] 0.1577783 0.1523971 0.3412915
density(ebola_si, at = 0.5)
#> [1] 0.6065307
quantile(ebola_si, p = 0.5)
#> [1] 0.6931472

ebola_si <- discretise(ebola_si)
ebola_si
#> Disease: ebola
#> Pathogen: NA
#> Epi Distribution: serial interval
#> Study: No citation available
#> Distribution: discrete gamma
#> Parameters:
#>   shape: 1.000
#>   scale: 1.000

cdf(ebola_si, q = 50)
#> [1] 1
generate(ebola_si, times = 10)
#>  [1] 2 2 2 1 1 1 2 1 2 5
density(ebola_si, at = 0.5)
#> [1] 0.3934693
quantile(ebola_si, p = 0.5)
#> [1] 0
```

<sup>Created on 2023-04-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

But I definitely agree we need to make this clear in the documentation and vignettes.

File: R/epidist.R (hunk 1 above)
Comment (adamkucharski) [reply R/epidist.R:434]: Thanks. Although I notice it seems to be converting (mean,dispersion) to (n,p) in epidist object here?

**Review by adamkucharski — APPROVED** [review]

File: R/epidist.R (hunk 1 above)
Comment (joshwlambert) [reply R/epidist.R:434]: Yes, that's due to the use of the {distributional} package underlying the {epiparameter} print method. We could write a conversion into the print method to make sure the parameters printed are the same as those input into the `<epidist>`, in this case mean and dispersion. 

What do you think?

**Pull request merged**